src/python/pants/backend/python/dependency_inference/import_parser.py

```diff
@@ -35,40 +34,55 @@ def all_imports(self) -> FrozenOrderedSet[str]:
         return FrozenOrderedSet(sorted([*self.explicit_imports, *self.inferred_imports]))
 
 
-def parse_file(*, filename: str, content: str) -> Optional[Tuple]:
+_Visitor = TypeVar("_Visitor")
+
+
+class VisitorInterface:
+    explicit_imports: Set[str]
+    inferred_imports: Set[str]
+
+    def __init__(self, module_name: str) -> None:
+        ...
+
+    @classmethod
+    def visit_tree(cls: Type[_Visitor], tree: Any, module_name: str) -> _Visitor:
+        ...
+
+
+def parse_file(*, filename: str, content: str, module_name: str) -> VisitorInterface:
+    """Parse the file for python imports, and return a visitor with the imports it found."""
     try:
-        # NB: The Python 3 ast is generally backwards-compatible with earlier versions. The only
-        # breaking change is `async` `await` becoming reserved keywords in Python 3.7 (deprecated
-        # in 3.6). If the std-lib fails to parse, we could use typed-ast to try parsing with a
-        # target version of Python 3.5, but we don't because Python 3.5 is almost EOL and has very
-        # low usage.
-        # We will also fail to parse Python 3.8 syntax if Pants is run with Python 3.6 or 3.7.
-        # There is no known workaround for this, beyond users changing their `./pants` script to
-        # always use >= 3.8.
-        tree = ast3.parse(content, filename=filename)
-        visitor_cls = _Py3AstVisitor if sys.version_info[:2] < (3, 8) else _Py38AstVisitor
-        return tree, visitor_cls
+        py27_tree = ast27.parse(content, filename=filename)
+        completed_visitor = _Py27AstVisitor.visit_tree(py27_tree, module_name=module_name)
+        return completed_visitor
     except SyntaxError:
-        try:
-            py27_tree = ast27.parse(content, filename=filename)
-            return py27_tree, _Py27AstVisitor
-        except SyntaxError:
-            return None
+        # NB: When the python 2 ast visitor fails to parse python 3 syntax, it raises a
+        # SyntaxError. This may also occur when the file contains invalid python code. If we parse a
+        # python 3 file with a python 2 parser, that should not change the imports we calculate.
+        pass
+
+    # Parse all python 3 code with libCST.
+    # NB: Since all python 3 code is forwards-compatible with the 3.8 parser, and the import syntax
+    # remains unchanged, we are safely able to use the 3.8 parser for parsing imports.
+    config = cst.PartialParserConfig(python_version="3.8")
+    cst_tree = cst.parse_module(content, config=config)
+    # NB: If the file contains invalid python code, it will raise an exception we expect to be
+    # caught.
```

Comment:
I'm not following this part. CST will not raise an exception on us for invalid syntax?

Reply:
It will! It raises a `cst.ParserSyntaxError` which we catch in `find_python_imports()`!